hpct-ldap-server/src/servers/ldap.py

```diff
@@ -0,0 +1,503 @@
+                "--generate-self-signed",
+                "--load-privkey",
+                "/etc/ssl/private/mycakey.pem",
+                "--template",
+                "/etc/ssl/ca.info",
+                "--outfile",
+                "/usr/local/share/ca-certificates/mycacert.crt",
+            ]
+
+            rc = subprocess.call(ca_cert_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create certificate authority.")
+
+            # Add CA certificate to trusted certs
+            rc = subprocess.call(
+                ["update-ca-certificates"], stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT
+            )
+
+            if rc != 0:
+                raise Exception("Unable to add CA certificate to trusted certs")
+
+            # Get Server Hostname
+            result = subprocess.run(["cat", "/etc/hostname"], capture_output=True, text=True)
+
+            # Private Key for Server
+            pks_args = [
+                "certtool",
+                "--generate-privkey",
+                "--bits",
+                "2048",
+                "--outfile",
+                "/etc/ldap/ldap01_slapd_key.pem",
+            ]
+            rc = subprocess.call(pks_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create private key for server.")
+
+            # Template for Server Certificate
+            sc_template = [
+                f"organization = {org}",
+                f"cn = {result.stdout}",
+                "tls_www_server",
+                "encryption_key",
+                "signing_key",
+                "expiration_days = 365",
+            ]
+            with open("/etc/ssl/ldap01.info", "w") as f:
+                for line in sc_template:
+                    f.write(f"{line}\n")
+
+            # Create Server Certificate
+            sc_args = [
+                "certtool",
+                "--generate-certificate",
+                "--load-privkey",
+                "/etc/ldap/ldap01_slapd_key.pem",
+                "--load-ca-certificate",
+                "/etc/ssl/certs/mycacert.pem",
+                "--load-ca-privkey",
+                "/etc/ssl/private/mycakey.pem",
+                "--template",
+                "/etc/ssl/ldap01.info",
+                "--outfile",
+                "/etc/ldap/ldap01_slapd_cert.pem",
+            ]
+            rc = subprocess.call(sc_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create server certificate.")
+
+            # Adjust Permissions and Ownership
+            g_args = ["chgrp", "openldap", "/etc/ldap/ldap01_slapd_key.pem"]
+            rc = subprocess.call(g_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to set group permissions.")
```

Comment:
This could use `os.chgrp`.

Reply:
Changed this to `shutil.chown`, couldn't find `os.chgrp`. I went with `shutil.chown` over `os.chown` since it accepts names not just numeric ids.

https://docs.python.org/3/library/os.html#os.chown